File: cpp/build-company/Warehouse.cpp
#include "Warehouse.hpp"
#include <iostream>
#include <algorithm>

// Реализация конструктора
Warehouse::Warehouse(const std::string& address) : address(address) {
}

// Реализация деструктора
Warehouse::~Warehouse() {
    std::cout << "Вызван деструктор Warehouse для: " << address << std::endl;
}

// Реализация геттеров
std::string Warehouse::getAddress() const {
    return address;
}

const std::vector<std::shared_ptr<Material>>& Warehouse::getMaterials() const {
    return materials;
}

const std::vector<std::shared_ptr<Equipment>>& Warehouse::getEquipment() const {
    return equipment;
}

// Реализация сеттера
void Warehouse::setAddress(const std::string& address) {
    this->address = address;
}

// Добавление материала
void Warehouse::addMaterial(std::shared_ptr<Material> material) {
    if (material) {
        materials.push_back(material);
        std::cout << "Добавлен материал '" << material->getName() << "' на склад по адресу " << address << std::endl;
    }
}

// Добавление оборудования
void Warehouse::addEquipment(std::shared_ptr<Equipment> equipment) {
    if (equipment) {
        this->equipment.push_back(equipment);
        std::cout << "Добавлено оборудование '" << equipment->getName() << "' на склад по адресу " << address << std::endl;
    }
}

// Удаление материала
void Warehouse::removeMaterial(std::shared_ptr<Material> material) {
    auto it = std::find(materials.begin(), materials.end(), material);
    if (it != materials.end()) {
        materials.erase(it);
        std::cout << "Удален материал '" << material->getName() << "' со склада" << std::endl;
    }
}

// Удаление оборудования
void Warehouse::removeEquipment(std::shared_ptr<Equipment> equipment) {
    auto it = std::find(this->equipment.begin(), this->equipment.end(), equipment);
    if (it != this->equipment.end()) {
        this->equipment.erase(it);
        std::cout << "Удалено оборудование '" << equipment->getName() << "' со склада" << std::endl;
    }
}

// Реализация функции отображения информации
void Warehouse::displayInfo() const {
    std::cout << "=== Склад ===" << std::endl;
    std::cout << "Адрес: " << address << std::endl;
    std::cout << "Всего материалов: " << materials.size() << std::endl;
    std::cout << "Всего оборудования: " << equipment.size() << std::endl;
}

// Перегрузка оператора [] для доступа к материалам по индексу
std::shared_ptr<Material> Warehouse::operator[](size_t index) {
    if (index < materials.size()) {
        return materials[index];
    }
    std::cout << "Ошибка: Индекс материала вне диапазона!" << std::endl;
    return nullptr;
}

const std::shared_ptr<Material> Warehouse::operator[](size_t index) const {
    if (index < materials.size()) {
        return materials[index];
    }
    std::cout << "Ошибка: Индекс материала вне диапазона!" << std::endl;
    return nullptr;
}

// Отображение полного инвентаря
void Warehouse::displayInventory() const {
    std::cout << "\n=== Инвентарь склада по адресу " << address << " ===" << std::endl;

    std::cout << "\n--- Материалы ---" << std::endl;
    if (materials.empty()) {
        std::cout << "Материалы отсутствуют" << std::endl;
    } else {
        for (const auto& material : materials) {
            material->displayInfo();
            std::cout << std::endl;
        }
    }

    std::cout << "--- Оборудование ---" << std::endl;
    if (equipment.empty()) {
        std::cout << "Оборудование отсутствует" << std::endl;
    } else {
        for (const auto& equip : equipment) {
            equip->displayInfo();
            std::cout << std::endl;
        }
    }
}

File: csharp/build-company/Warehouse.cs
namespace BuildCompany;

/// <summary>
/// Класс, представляющий склад для материалов и оборудования
/// </summary>
public class Warehouse
{
    private string address;
    private List<Material> materials;
    private List<Equipment> equipment;

    /// <summary>
    /// Конструктор
    /// </summary>
    /// <param name="address">Адрес склада</param>
    public Warehouse(string address)
    {
        this.address = address;
        this.materials = new List<Material>();
        this.equipment = new List<Equipment>();
    }

    /// <summary>
    /// Получить адрес
    /// </summary>
    public string GetAddress()
    {
        return address;
    }

    /// <summary>
    /// Получить список материалов
    /// </summary>
    public List<Material> GetMaterials()
    {
        return materials;
    }

    /// <summary>
    /// Получить список оборудования
    /// </summary>
    public List<Equipment> GetEquipment()
    {
        return equipment;
    }

    /// <summary>
    /// Установить адрес
    /// </summary>
    public void SetAddress(string address)
    {
        this.address = address;
    }

    /// <summary>
    /// Добавить материал
    /// </summary>
    public void AddMaterial(Material material)
    {
        materials.Add(material);
        Console.WriteLine($"Material '{material.GetName()}' added to warehouse at {address}");
    }

    /// <summary>
    /// Добавить оборудование
    /// </summary>
    public void AddEquipment(Equipment equip)
    {
        equipment.Add(equip);
        Console.WriteLine($"Equipment '{equip.GetName()}' added to warehouse at {address}");
    }

    /// <summary>
    /// Удалить материал
    /// </summary>
    public void RemoveMaterial(Material material)
    {
        materials.Remove(material);
        Console.WriteLine($"Material '{material.GetName()}' removed from warehouse");
    }

    /// <summary>
    /// Удалить оборудование
    /// </summary>
    public void RemoveEquipment(Equipment equip)
    {
        equipment.Remove(equip);
        Console.WriteLine($"Equipment '{equip.GetName()}' removed from warehouse");
    }

    /// <summary>
    /// Отображение информации о складе
    /// </summary>
    public void DisplayInfo()
    {
        Console.WriteLine($"[Warehouse]");
        Console.WriteLine($"  Address: {address}");
        Console.WriteLine($"  Materials stored: {materials.Count}");
        Console.WriteLine($"  Equipment stored: {equipment.Count}");
    }

    /// <summary>
    /// Отображение инвентаря склада
    /// </summary>
    public void DisplayInventory()
    {
        Console.WriteLine($"\n=== Warehouse Inventory at {address} ===");

        Console.WriteLine($"\nMaterials ({materials.Count}):");
        foreach (var material in materials)
        {
            material.DisplayInfo();
            Console.WriteLine();
        }

        Console.WriteLine($"\nEquipment ({equipment.Count}):");
        foreach (var equip in equipment)
        {
            equip.DisplayInfo();
            Console.WriteLine();
        }
    }
}
